### ISSUE TITLE
TestRef example is missleading and pointless.

### ISSUE THREAD

**Issue by Alumniminium** [dotnet/docs#7416]
arrays are reference types, passing them with `ref ` changes nothing. You can remove the `ref ` from `static void FillArray(ref int[] arr)` and from `FillArray(ref theArray);` and the result will be the same. This is very very confusing to people trying to learn from the official docs.

With and without using `ref ` the following output is achieved:

**Output:** 
Array elements are:
1111 2 3 4 5555 Press any key to exit.
---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: e4e7c322-b662-56f9-99af-f47106af0780
* Version Independent ID: ae1843c3-6606-c1fa-50c8-f1ac99ad5d05
* Content: [Passing Arrays Using ref and out (C# Programming Guide)](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/arrays/passing-arrays-using-ref-and-out)
* Content Source: [docs/csharp/programming-guide/arrays/passing-arrays-using-ref-and-out.md](https://github.com/dotnet/docs/blob/master/docs/csharp/programming-guide/arrays/passing-arrays-using-ref-and-out.md)
* Product: **dotnet-csharp**
* GitHub Login: @BillWagner
* Microsoft Alias: **wiwagn**

**Comment by pkulikov** [comment]
@Alumniminium 
>arrays are reference types, passing them with `ref` changes nothing.

passing with `ref` changes a lot :) if you pass with `ref` you can change to **what** reference references. Consider the following example code (I've already put outputs into comments, however you might just copy it and run in [try.dot.net](https://try.dot.net/)):

```csharp
using System;

public class Program
{
  public static void Main()
  {
    int[] origin = { 1, 2, 3, 4, 5 };
    int[] copy = origin;
    ClearArray(origin);
    Display(origin);  // 0 0 0 0 0
    Display(copy);    // 0 0 0 0 0

    int[] origin2 = { 1, 2, 3, 4, 5 };
    int[] copy2 = origin2;
    ClearArrayRef(ref origin2);
    Display(origin2);  // 0 0 0 0 0
    Display(copy2);    // 1 2 3 4 5  remark: copy2 and origin2 reference two different arrays

    int[] origin3 = { 1, 2, 3, 4, 5 };
    int[] copy3 = origin3;
    ClearArrayAsIfRef(origin3);
    Display(origin3);  // 1 2 3 4 5   remark: method has failed to clear array.
    Display(copy3);    // 1 2 3 4 5
  }

  private static void Display(int[] arr) => Console.WriteLine(string.Join(" ", arr));

  private static void ClearArray(int[] input)
  {
    for (var i = 0; i < input.Length; i++)
    {
      input[i] = 0;
    }
  }

  private static void ClearArrayRef(ref int[] input)
  {
    input = new int[input.Length];
    for (var i = 0; i < input.Length; i++)
    {
      input[i] = 0;
    }
  }

  private static void ClearArrayAsIfRef(int[] input)
  {
    input = new int[input.Length];
    for (var i = 0; i < input.Length; i++)
    {
      input[i] = 0;
    }
  }
}
```

Note that `ClearArrayRef` and `ClearArrayAsIfRef` differ only by `ref`, however, they behave very differently.

I agree that documentation should provide better examples that would highlight the difference.

**Comment by BillWagner** [comment]
I had some internal discussions, and here's our recommendation:

This article should be removed, and it should be redirected to the [array index page](https://docs.microsoft.com/dotnet/csharp/programming-guide/arrays/).  

@Alumniminium is correct that this is a misleading set of examples. @pkulikov is also correct that there are semantic differences in passing arrays (or any object) by reference. We tried to formulate a good example that would pass an array by reference. Any example we constructed could be a better API design using a combination of return values, or `out` arguments.

I've added this to the backlog, and added the up-for-grabs label, if anyone wants to submit a PR for the fix.
